core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -366,18 +368,21 @@ final class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: Stri
     }
 
     // traverse the list and try to complete some watched elements
-    def tryCompleteWatched(): Int = {
+    def tryCompleteWatched(stopEarly: Boolean): Int = {
       var completed = 0
 
       val iter = operations.iterator()
-      while (iter.hasNext) {
+      var couldStop = false
+      while (iter.hasNext && !couldStop) {
```

Comment:
Hmm... `operations` is a queue. How is it keeping the expected ordering?

Reply:
We are assuming that the put calls would be in ascending order of the completion threshold and hence we can stop early, which is the case for raft.

Reply:
Seems like that relies on some brittle assumptions. For example, truncation could result in non-monotonic offsets. Also, I don't think it could be used on the fetch side where offsets would not be added in any specific order, which is what I was hoping to do in #377.

Reply:
For fetching currently I'm using constant value 0 so that it is still monotonic; but of course if we change it to offsets then it is not monotonic anymore, but I guess I can make the fetch / produce purgatory differently so that the latter can still exit early to avoid always traversing through the whole watch list.

Reply:
I think in practice there won't be _that_ many pending operations, so maybe it's reasonable to iterate all of them for now. We can always come up with a more efficient data structure later.

Reply:
For produce purgatory we may actually have many pending operations if the diff between log-end to hw is large.

Let me first try to refactor the fetch purgatory based on offset value and see if I feel comfortable reverting for produce purgatory.

Reply:
I'm still not too thrilled about this optimization. Could we leave this for future work if/when we have some profiling which shows that it is really worthwhile?